addons/product/models/product_template.py

```diff
@@ -848,63 +862,81 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
         :param necessary_values: values that must be in the returned combination
         :type necessary_values: recordset of `product.template.attribute.value`
 
-        :return: the first possible combination found, or empty if none possible
-        :rtype: recordset of `product.template.attribute.value`
+        :return: the possible combinations
+        :rtype: generator of recordset of `product.template.attribute.value`
         """
         self.ensure_one()
 
         if not self.active:
-            return self.env['product.template.attribute.value']
+            return _("The product template is archived so no combination is possible.")
 
-        attribute_lines = self._get_valid_product_template_attribute_lines()
+        ptal_stack = [self._get_valid_product_template_attribute_lines()]
```

Comment:
recordset are iterable by default

Reply:
This is a stack of recordset (with potentially multiple values in each).

[0] = all attributes lines
[1] = all attributes lines - 1
[2] = all attributes lines - 2
etc until it's empty which is the stop condition for the loop.

The combination_stack is doing the reverse.
[0] = no value
[1] = value 2 of line n-1
[2] = value 1 of line n-1
etc